samples/matter/lock/src/credentials/credentials_data_types.cpp

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include "credentials_data_types.h"
+
+namespace DoorLockData
+{
+
+void pack(void *buff, const void *data, size_t dataSize, size_t &offset)
+{
+	memcpy(reinterpret_cast<uint8_t *>(buff) + offset, data, dataSize);
+	offset += dataSize;
+}
+
+void unpack(void *data, size_t dataSize, const void *buff, size_t &offset)
+{
+	memcpy(data, reinterpret_cast<const uint8_t *>(buff) + offset, dataSize);
+	offset += dataSize;
+}
+
+CHIP_ERROR Credential::FillFromPlugin(EmberAfPluginDoorLockCredentialInfo &credentialInfo)
+{
+	if (credentialInfo.credentialData.size() < kMaxCredentialLength) {
+		mInfo.mFields.mStatus = static_cast<uint8_t>(credentialInfo.status);
+		mInfo.mFields.mCredentialType = static_cast<uint8_t>(credentialInfo.credentialType);
+		mInfo.mFields.mCreationSource = static_cast<uint8_t>(credentialInfo.creationSource);
+		mInfo.mFields.mCreatedBy = static_cast<uint8_t>(credentialInfo.createdBy);
+		mInfo.mFields.mModificationSource = static_cast<uint8_t>(credentialInfo.modificationSource);
+		mInfo.mFields.mLastModifiedBy = static_cast<uint8_t>(credentialInfo.lastModifiedBy);
+		mSecret.mDataLength = credentialInfo.credentialData.size();
+		memcpy(mSecret.mData, credentialInfo.credentialData.data(), credentialInfo.credentialData.size());
+		return CHIP_NO_ERROR;
+	}
+	return CHIP_ERROR_BUFFER_TOO_SMALL;
+}
+
+CHIP_ERROR Credential::ConvertToPlugin(EmberAfPluginDoorLockCredentialInfo &credentialInfo) const
+{
+	if (mSecret.mDataLength < RequiredBufferSize()) {
+		credentialInfo.status = static_cast<DlCredentialStatus>(mInfo.mFields.mStatus);
+		credentialInfo.credentialType = static_cast<CredentialTypeEnum>(mInfo.mFields.mCredentialType);
+		credentialInfo.credentialData = chip::ByteSpan(mSecret.mData, mSecret.mDataLength);
+		credentialInfo.creationSource = static_cast<DlAssetSource>(mInfo.mFields.mCreationSource);
+		credentialInfo.createdBy = static_cast<chip::FabricIndex>(mInfo.mFields.mCreatedBy);
+		credentialInfo.modificationSource = static_cast<DlAssetSource>(mInfo.mFields.mModificationSource);
+		credentialInfo.lastModifiedBy = static_cast<chip::FabricIndex>(mInfo.mFields.mLastModifiedBy);
+		return CHIP_NO_ERROR;
+	}
+	return CHIP_ERROR_BUFFER_TOO_SMALL;
+}
+
+size_t Credential::Serialize(void *buff, size_t buffSize)
+{
+	if (!buff || buffSize < RequiredBufferSize()) {
+		return 0;
+	}
+	size_t offset = 0;
+
+	pack(buff, mInfo.mRaw, sizeof(mInfo.mRaw), offset);
+	pack(buff, &mSecret.mDataLength, sizeof(mSecret.mDataLength), offset);
+	pack(buff, mSecret.mData, mSecret.mDataLength, offset);
+
+	return offset;
+}
+
+CHIP_ERROR Credential::Deserialize(const void *buff, size_t buffSize)
+{
+	if (!buff) {
+		return CHIP_ERROR_INVALID_ARGUMENT;
+	}
+
+	if (buffSize > RequiredBufferSize()) {
```

Comment:
nit: you should probably also check if buffSize is not less than `sizeof(mInfo.mRaw) + sizeof(mSecret.mDataLength)`, but this will be solved automatically after switching to `BufferReader`.

Reply:
Yes, that's a valid point but on the other hand the API documentation clearly states that "buffSize` must be set to at least the RequiredBufferSize size." when using Serialize() methods :)

Reply:
On the other hand this does not fit into the `buffSize > RequiredBufferSize()` check :) I guess this should be revisited during the refactoring task that is already scheduled.